File: nodejs/backend/src/dto/leadRequestDto.ts
import { z } from 'zod';
import { LeadRequest } from '../model/LeadRequest';

export const createLeadRequestSchema = z.object({
  name: z.string().trim().min(1, 'name must not be empty').max(220),
  email: z.string().trim().email('must be a valid email address'),
  phone: z.string().trim().min(1, 'phone must not be empty').max(15),
  service: z.string().trim().min(1, 'service must not be empty').max(520),
  message: z.string().max(2000, 'message is too long'),
});
export type CreateLeadRequestDto = z.infer<typeof createLeadRequestSchema>;

export interface LeadRequestResponseDto {
  id: string;
  name: string;
  email: string;
  phone: string;
  service: string;
  message: string;
  createdAt: Date;
}

export function toLeadRequestResponseDto(r: LeadRequest): LeadRequestResponseDto {
  return {
    id: r.id,
    name: r.name,
    email: r.email,
    phone: r.phone,
    service: r.service,
    message: r.message,
    createdAt: r.createdAt,
  };
}

File: rust/backend/src/dto/lead_request_dto.rs
use crate::model::LeadRequest;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use validator::Validate;

#[derive(Debug, Deserialize, Validate)]
pub struct CreateLeadRequest {
    #[validate(length(min = 1, max = 120, message = "name must not be empty"))]
    pub name: String,

    #[validate(email(message = "must be a valid email address"))]
    pub email: String,

    #[validate(length(min = 1, max = 30, message = "phone must not be empty"))]
    pub phone: String,

    #[validate(length(min = 1, max = 120, message = "service must not be empty"))]
    pub service: String,

    #[validate(length(max = 2000, message = "message is too long"))]
    pub message: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct LeadRequestResponse {
    pub id: Uuid,
    pub name: String,
    pub email: String,
    pub phone: String,
    pub service: String,
    pub message: String,
    pub created_at: DateTime<Utc>,
}

impl From<LeadRequest> for LeadRequestResponse {
    fn from(r: LeadRequest) -> Self {
        Self {
            id: r.id,
            name: r.name,
            email: r.email,
            phone: r.phone,
            service: r.service,
            message: r.message,
            created_at: r.created_at,
        }
    }
}
